src/ggrc/assets/javascripts/components/access_control_list/access_control_list.js

```diff
@@ -0,0 +1,212 @@
+          inst.attr('access_control_list'),
+          {person_id: person.id, ac_role_id: roleId}
+        );
+
+        if (idx < 0) {
+          console.warn('Role ', roleId, 'not found for user', person.id);
+          return;
+        }
+
+        inst.access_control_list.splice(idx, 1);
+
+        if (!this.autosave) {
+          this._rebuildRolesInfo();
+          return;
+        }
+
+        inst.save()
+          .done(function () {
+            GGRC.Errors.notifier('success', 'User role removed.');
+            this._rebuildRolesInfo();
+          }.bind(this))
+          .fail(function () {
+            GGRC.Errors.notifier('error', 'Removing user role failed.');
+          });
+      },
+
+      /**
+       * Update the role info object used by the template to display the roles
+       * and the users which have them granted.
+       */
+      _rebuildRolesInfo: function () {
+        var roleAssignments;
+        var roles;
+        var rolesInfo;
+        var instance = this.instance;
+
+        if (!instance) {
+          this.attr('rolesInfo', []);
+          return;
+        }
+
+        roleAssignments = _.groupBy(instance.access_control_list, 'ac_role_id');
+
+        roles = _.filter(
+          GGRC.access_control_roles, {object_type: instance.type});
+
+        rolesInfo = _.map(roles, function (role) {
+          return {role: role, roleAssignments: roleAssignments[role.id]};
+        });
+
+        this.attr('rolesInfo', rolesInfo);
+      }
+    },
+
+    events: {
+      /**
+       * The component entry point.
+       *
+       * @param {jQuery} $element - the DOM element that triggered
+       *   creating the component instance
+       * @param {Object} options - the component instantiation options
+       */
+      init: function ($element, options) {
+        var vm = this.viewModel;
+
+        vm.attr('grantingRoleId', 0);
+        vm._rebuildRolesInfo();
+      },
+
+      '{viewModel.instance.access_control_list} change':
+      function (context, ev, index, how, newVal, oldVal) {
+        if (how === 'set') {
+          // we are not interested in collection items' changes, all we care
+          // about is ading and removing role assignments
+          return;
+        }
+
+        this.viewModel._rebuildRolesInfo();
+      }
+
+      //////////////////////////////////
+      // TODO: manually correcting broken object is needed for some reson,
+      // have a closer look!
+      , '{viewModel.rolesInfo} change': function () {
+        var vm = this.viewModel;
+
+        console.info('checking rolesInfo correctness');
+        vm.rolesInfo.forEach(function (item, i, array) {
+          var roleId = item.role.id;
+          if (!item.roleAssignments) {
+            return;
+          }
+          item.roleAssignments.forEach(function (ass) {
+            if (ass.ac_role_id !== roleId) {
+              console.warn('ERROR 333, rebuilding rolesInfo');
+              vm._rebuildRolesInfo();
```

Comment:
Try to comment out this line, and you might see people listed under an incorrect role (e.g. users with Role X listed under the role Y).

This is really weird, because it seems that something corrupts the viewModel's role info object from the outside - performing the validation check on the object's contents each time it is rebuilt in the component (i.e. in the `_rebuldRolesInfo()` method) did not detect any errors - but something still changes it.

If jumping into debug mode when the data corruption is detected, it can be seen in the Network tab that there is a pending people GET request - if that matters.

The way I was able to reproduce the issue was by having a Control instance with 3 custom roles -  one user assigned to the first role, two (different) users assigned to the second role and no users assigned to the third role.

Interestingly, this issue happens only once per page load. Opening an Edit modal, switching to a different Control in tree view and back to the problematic control etc., all these actions make the problem disappear and debug IDs on the Control info pane all match.